src/engine.cpp

```diff
@@ -37,43 +37,89 @@ using namespace asio::ip;
 
 using namespace cocaine;
 
-execution_unit_t::execution_unit_t(context_t& context):
-    m_asio(new io_service()),
-    m_chamber(new io::chamber_t("core:asio", m_asio))
+class execution_unit_t::gc_action_t:
+    public std::enable_shared_from_this<gc_action_t>
 {
-    m_log = context.log("core:asio", {
-        attribute::make("engine", boost::lexical_cast<std::string>(m_chamber->thread_id()))
-    });
+    execution_unit_t *const parent;
+    const boost::posix_time::seconds repeat;
 
-    COCAINE_LOG_DEBUG(m_log, "engine started");
-}
+public:
+    template<class Interval>
+    gc_action_t(execution_unit_t *const parent_, Interval repeat_):
+        parent(parent_),
+        repeat(repeat_)
+    { }
 
-namespace {
-
-struct detach_action_t {
     void
     operator()();
 
-    // Never becomes a dangling reference.
-    std::map<int, std::shared_ptr<session_t>>& sessions;
+private:
+    void
+    finalize(const std::error_code& ec);
 };
 
 void
-detach_action_t::operator()() {
-    for(auto it = sessions.begin(); it != sessions.end(); ++it) {
-        it->second->detach();
+execution_unit_t::gc_action_t::operator()() {
+    parent->m_cron.expires_from_now(repeat);
+
+    parent->m_cron.async_wait(std::bind(&gc_action_t::finalize,
+        shared_from_this(),
+        std::placeholders::_1
+    ));
+}
+
+void
+execution_unit_t::gc_action_t::finalize(const std::error_code& ec) {
+    if(ec == asio::error::operation_aborted) {
+        return;
+    }
+
+    size_t recycled = 0;
+
+    for(auto it = parent->m_sessions.begin(); it != parent->m_sessions.end();) {
+        if(!it->second->memory_pressure()) {
+            recycled++;
+            it = parent->m_sessions.erase(it);
+            continue;
+        }
+
+        ++it;
     }
+
+    if(recycled) {
+        COCAINE_LOG_DEBUG(parent->m_log, "recycled %llu session(s)", recycled);
+    }
+
+    operator()();
 }
 
-} // namespace
+execution_unit_t::execution_unit_t(context_t& context):
+    m_asio(new io_service()),
+    m_chamber(new io::chamber_t("core:asio", m_asio)),
+    m_cron(*m_asio)
+{
+    m_log = context.log("core:asio", {
+        attribute::make("engine", boost::lexical_cast<std::string>(m_chamber->thread_id()))
+    });
+
+    m_asio->post(std::bind(&gc_action_t::operator(),
+        std::make_shared<gc_action_t>(this, boost::posix_time::seconds(kCollectionInterval))
+    ));
+
+    COCAINE_LOG_DEBUG(m_log, "engine started");
+}
 
 execution_unit_t::~execution_unit_t() {
-    if(!m_sessions.empty()) {
-        COCAINE_LOG_DEBUG(m_log, "engine waiting for outstanding operations to complete");
+    m_asio->post([this] {
+        COCAINE_LOG_DEBUG(m_log, "stopping engine");
 
-        // Asynchronously close the connections.
-        m_asio->post(detach_action_t{m_sessions});
-    }
+        for(auto it = m_sessions.begin(); it != m_sessions.end(); ++it) {
+            // Close the connections.
+            it->second->detach(std::error_code());
```

Comment:
I'm actually not sure about the error code here. It's not really an ERROR when we shutdown the engine normally as a part of runtime shutdown, right? Also, no operation was aborted, technically.